icu4j/main/classes/core/src/com/ibm/icu/util/ULocale.java

```diff
@@ -1203,6 +1205,328 @@ public static String getKeywordValue(String localeID, String keywordName) {
+         *
+         * If the alias data has already loaded, then this method simply
+         * returns without doing anything meaningful.
+         *
+         */
+        private static synchronized void loadAliasData() {
+            if (aliasDataIsLoaded) {
+                return;
+            }
+            languageAliasMap = new HashMap<String, String>();
+            scriptAliasMap = new HashMap<String, String>();
+            territoryAliasMap = new HashMap<String, List<String>>();
+            variantAliasMap = new HashMap<String, String>();
+
+            UResourceBundle metadata = UResourceBundle.getBundleInstance(
+                ICUData.ICU_BASE_NAME, "metadata",
+                ICUResourceBundle.ICU_DATA_CLASS_LOADER);
+            UResourceBundle metadataAlias = metadata.get("alias");
+            UResourceBundle languageAlias = metadataAlias.get("language");
+            UResourceBundle scriptAlias = metadataAlias.get("script");
+            UResourceBundle territoryAlias = metadataAlias.get("territory");
+            UResourceBundle variantAlias = metadataAlias.get("variant");
+
+            for ( int i = 0 ; i < languageAlias.getSize(); i++ ) {
+                UResourceBundle res = languageAlias.get(i);
+                String aliasFrom = res.getKey();
+                String aliasTo = res.get("replacement").getString();
+                languageAliasMap.put(aliasFrom, aliasTo);
+            }
+            for ( int i = 0 ; i < scriptAlias.getSize(); i++ ) {
+                UResourceBundle res = scriptAlias.get(i);
+                String aliasFrom = res.getKey();
+                String aliasTo = res.get("replacement").getString();
+                scriptAliasMap.put(aliasFrom, aliasTo);
+            }
+            for ( int i = 0 ; i < territoryAlias.getSize(); i++ ) {
+                UResourceBundle res = territoryAlias.get(i);
+                String aliasFrom = res.getKey();
+                String aliasTo = res.get("replacement").getString();
+                territoryAliasMap.put(aliasFrom,
+                    new ArrayList<String>(Arrays.asList(aliasTo.split(" "))));
+            }
+            for ( int i = 0 ; i < variantAlias.getSize(); i++ ) {
+                UResourceBundle res = variantAlias.get(i);
+                String aliasFrom = res.getKey();
+                String aliasTo = res.get("replacement").getString();
+                variantAliasMap.put(aliasFrom, aliasTo);
+            }
+
+            aliasDataIsLoaded = true;
+        }
+
+        private static String generateKey(
+                String language, String region, String variant) {
+            StringBuilder buf = new StringBuilder();
+            if (language != null) {
+                buf.append(language);
+            }
+            if (region != null && !region.isEmpty()) {
+                buf.append(UNDERSCORE);
+                buf.append(region);
+            }
+            if (variant != null && !variant.isEmpty()) {
```

Comment:
well. I have to filter that out outside this function otherwise, the logic would mass up of the replacement, because the outter function may believe the  ill-formed variant is in the type but it is not in the type while we access the key. for example, if we have "nb_NO_NY", if we use >= 4 the outter function will think the type is nb_NY while the key generated is really nb only, then when it find the replacement for "no" for "nb", it will see there are no variant in the replacement but believe the NY is in the type and therefore remove the NY . So if we do not want to search for nb_NY, I have to ignore the NY outside this function so the outter function think the key is only "nb" and therefore won't try to remove "NY" from it.